docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+
+For example, the [`Schema`](#schema) object contains a [`files`](#files) property, which is a list of `DescFile` objects representing the files requested to be generated.  The first thing you will most likely do in your generator function is iterate over this list and issue a call to the [`generateFile`](#generatefile) function for each file object.  This will return a generated file object containing a `print` function which you can then use to "print" to the file.
+
+Each `DescFile` object in the [`files`](#files) property contains all enums and messages needed to begin generating code:
+
+```ts
+// protoc-gen-foo-plugin.js
+
+function generateTs(schema: Schema) {
+   for (const file of schema.files) {
+     const f = schema.generateFile(file.name + "_pb.ts");
+  
+     for (const enumeration of file.enums) {
+        f.print(`// generating enums from ${file.name}`);
+        ...
+     }
+  
+     for (const message of file.messages) {
+        f.print(`// generating messages from ${file.name}`);
+        ...
+     }
+   }
+}
+```
+
+#### Reading Custom Options
+
+Protobuf-ES does not yet provide support for extensions, neither in general as pertaining to proto2 nor with custom options in proto3.  However, in the interim, there are convenience functions for retrieving any custom options specified in your .proto files.  These are provided as a temporary utility until full extension support is implemented.
+
+For available functions, see the [custom options](#custom-options) section under **API**.
+
+
+### Initializing your plugin
+
+Once your generator functions are defined, you then need to initialize your plugin by calling `createEcmaScriptPlugin` with a plugin initialization object.  
+
+```ts
+// protoc-gen-foo-plugin.js
+
+export const protocGenFoo = createEcmaScriptPlugin({
+   name: "protoc-gen-foo",
+   version: "v0.1.0",
+   generateTs,
+});
+```
+
+See the [`PluginInit`](#plugininit) object definition for more information.
+
+### Create an executable file
+
+Then, you need to create an executable file that will execute the plugin returned from the above call:
+
+```js
+// bin/protoc-gen-foo
+
+#!/usr/bin/env node
+
+const { runNodeJs } = require("@bufbuild/protoplugin");
+const { protocGenFoo } = require("../protoc-gen-foo-plugin.js");
+
+runNodeJs(protocGenFoo);
+```
+
+### Specify the executable
+
+Finally, specify the path to the executable file in the `bin` property of your `package.json` file:
+
+```js
+// package.json
+
+{
+  "name": "protoc-gen-foo",
+  "version": "0.1.0",
+  "bin": {
+    "protoc-gen-foo": "bin/protoc-gen-foo"
+  },
+  ...
+}
+
+```
+
+For a complete working example of a plugin written with the framework, check out [protoc-gen-es](https://github.com/bufbuild/protobuf-es/tree/main/packages/protoc-gen-es).
+
+
+## API
+
+Following are the definitions of relevant objects in the plugin framework.
```

Comment:
I don't think we should try to provide an API doc here (in the sense of a document that aims to list all functions, parameters, etc), because it takes too much effort to maintain without tooling like [jsdoc](https://jsdoc.app/about-getting-started.html#generating-a-website) or [typedoc](https://typedoc.org/).

Instead, I think should use a conversational style, similar to a tutorial, and explain/give examples for common topics in separate sections. For reference, take a look at the [runtime API doc for protobuf-es](https://github.com/bufbuild/protobuf-es/blob/sayers/plugin_docs/docs/runtime_api.md).

For @bufbuild/protoplugin, the list of sections could be:

- Setting up a plugin
  - Why give `version` and `name`? (part of preamble in generated code, shown with `protoc-gen-foo -h`)
- Providing a generator function
- How to walk through the Schema
- Generating a file (naming convention! we want users to add `_foo` for `protoc-gen-foo`)
- Printing to a generated file (at this point, we need to explain how to run the plugin)
- Importing
  - from a npm package
  - from `protoc-gen-es` generated code
- Exporting symbols, and importing them in other generated files
- Parsing plugin options
- Accessing custom protobuf options
- Generating JavaScript or TypeScript declarations yourself (explaining why this can be worth the extra effort)

It would also be great if we could provide answers for the questions:

- How to test my plugin?
- How can I generator non-ECMAScript files?

Reply:
Tried to implement all of this except the last two (testing and non-ECMAScript).  I wasn't entirely sure of the best way to do this yet.  I still need to try out generating a non-ES file.  For testing, I was thinking of pointing users to our protoplugin-test package, but not sure if that helps them.